design/kiota-e2e.md

```diff
@@ -0,0 +1,215 @@
+
+Models are also available in this package and should reflect the underlying version of Graph we are targeting. Developers using our types should do it in a way that will be familiar to how they are used to with other SDKs and APIs. We should provide an easy-to-use model that feels natural in the Javascript world. This means we should not be forcing developers to use Classes and / or other structures that will make their code less natural to them.
+
+We have an open issue about the above topic that can be followed here : [TypeScript - Use of interface models instead of class models](https://github.com/microsoft/kiota/issues/1013)
+
+```typescript
+const me = await getMe();
+const meRaw = await getMeRaw();
+await sendMail();
+await sendMailRaw();
+
+// The types returned by the fluent API should be the same as the .api() area. It should also be the same (or at least very similar) as the current @microsoft/microsoft-graph-types to offer seamless migration.
+async function getMe(): Promise<User | undefined> {
+	return await graphClient.me.get();
+}
+
+// Allowing raw calls (using the .api() method instead of the full fluent API) is important for migration purposes and cases we don't know the resource beforehands (thinking Graph Explorer, mgt-get, etc.)
+async function getMeRaw(): Promise<User | undefined> {
+	return await graphClient.api("/me").get();
+}
+
+// Sending an email via the fluent API
+async function sendMail(): Promise<void> {
+	const message: Message = {
+		subject: "Hello Graph TypeScript SDK!",
+		body: {
+			contentType: BodyType.Html,
+			content: "<bold>Hello Graph TypeScript SDK!</bold>",
+		},
+		toRecipients: [
+			{
+				emailAddress: {
+					address: "admin@m365x263716.onmicrosoft.com",
+				},
+			},
+		],
+	};
+
+	return await client.me.sendMail.post(message);
+}
+
+// Sending the email via the .api() method
+async function sendMailRaw(): Promise<void> {
+	const message: Message = {
+		subject: "Hello Graph TypeScript SDK!",
+		body: {
+			contentType: BodyType.Html,
+			content: "<bold>Hello Graph TypeScript SDK!</bold>",
+		},
+		toRecipients: [
+			{
+				emailAddress: {
+					address: "admin@m365x263716.onmicrosoft.com",
+				},
+			},
+		],
+	};
+
+	return await client.api("/me/sendMail").post({
+		message: message,
+	});
+}
+```
+
+## NodeJS e2e using the Core library
+
+### Concept
+
+This example brings very similar concepts, but utilizing only the core library. That means that as a developer, we only care about the core capabilities (authentication, serialization, etc.) but we don't need / want the fluent API. This also highlights that types are still important in this case (similarly to how they were important with the v3 of the SDK using our Typings).
```

Comment:
Since core functionality is also available in the service library, why do we want to expose the core library as a separate package? The problem I see with that is that it offers developers another choice that we'll need to explain so that developers can understand when to use which library and what are the trade-offs and limitations of each package. It could also lead to confusion if developers install the core library but want to try to use the fluent API and are stuck because it won't work, not knowing that for that they need a different package. So rather than offering two separate packages, I'd suggest that we offer just the service library with the fluent API being the recommended way of calling Graph. If need be, we offer developers to use `.api()` as an escape hatch, but both options are already available in the same package. As developers gradually upgrade their app from the raw to fluent API to benefit from the design-time typesafety features, they can do so without any additional dependencies.

Reply:
service libraries will be a layer on top of the core lib, which will offer the api() method like it does today. So effectively the consumers only will need to install the service lib they want to use, plus the side-car for authentication depending on their "platform" and they'll see everything, the api method and the fluent style APIs.

Reply:
Understood, so theoretically we could choose to expose a single package (service library that does both fluent and raw API) and keep the core library internal and not something we'd ever mention to developers, right?

Reply:
I don't know if we would keep it internal, but definitely not publicize its use. We believe having the core is critical to some advanced scenarios. When using the service library (which will be the preferred and documented approach), you would not need to add the core library explicitely.

Reply:
correct, to the exception of the developers who don't want the fluent stuff AT ALL, and just want the core lib. This population would install only core. But that's a niche segment, we'll have documentation for that, but we won't be advertising for it.

Reply:
@sebastienlevert  I suggest we should decide on the publicity of the core and service library based on the usage of the audience and this can be different for different languages. JS being a scripting language is used by a  wide range of users including non-developers. It would be cool to present the libraries as options that the user can pick based on their usage and comfort level.

Reply:
> It would be cool to present the libraries as options that the user can pick based on their usage and comfort level.

If you look across all of Microsoft 365, we already ask people to choose between many UI libraries, SDKs, types of apps, toolchains and tools. Adding yet another choice offers additional flexibility, but also requires us to properly explain its ramifications, constraints, recommendations, samples, tutorials and guidance. It also increases the risk of people choosing the wrong thing and having to back track if they find it's not what they need. We should consider this when deciding how to expose the SDKs.

Reply:
I absolutely 100% agree with @waldekmastykarz here. Our core is a tool we use to avoid duplication of code and allow common features to be shared with v1 and beta. I would argue that we should not publicize the core libraries (except for very advanced scenarios) as they are getting the exact same value via the service library. Our tree-shaking should deliver what the user needs, and this should be delivered via our service library.